### ISSUE TITLE
Failed to convert tf pb file (MTCNN model) to tflite format

### ISSUE THREAD

**Issue by kangshan0910** [tensorflow/tensorflow#16091]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04.2 LTS
- **TensorFlow installed from (source or binary)**: pip
- **TensorFlow version (use command below)**: 1.4.1
- **Python version**: 2.7.12
- **Bazel version (if compiling from source)**: 0.9.0
- **GCC/Compiler version (if compiling from source)**:gcc (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609
- **CUDA/cuDNN version**: n/a
- **GPU model and memory**:n/a
- **Exact command to reproduce**: bazel run --config=opt --copt=-msse4.1 --copt=-msse4.2  //tensorflow/contrib/lite/toco:toco --  --input_file=/home/xxx/facenet/ks/onet.pb   --output_file=/home/xxx/facenet/ks/onet.tflite  --input_format=TENSORFLOW_GRAPHDEF   --output_format=TFLITE   --inference_type=FLOAT   --input_shape=1,48,48,3   --input_array=onet/input --output_array=onet/prob1

### Describe the problem
I failed to covert my pb file to tflite format because some operations were not supported. Is there any workaround for the issue? 

### Source code / logs
2018-01-13 09:47:41.439557: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.439661: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.439687: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.439736: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.439767: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.439786: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.439824: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.439864: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.439884: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.439940: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.439970: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.439987: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.440027: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.440055: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.440073: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.440107: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.440135: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.440152: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.440208: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.440240: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.440259: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.440302: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.440333: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.440352: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.440392: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.440423: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.440442: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.440500: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.440533: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.440552: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.440596: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.440630: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.440650: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.440689: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.440720: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.440739: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.440798: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.440828: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.440851: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.440886: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.440918: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.440937: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.440987: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.441018: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.441037: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.441090: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.441118: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.441135: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.441168: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.441196: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.441213: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.441254: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Exp
2018-01-13 09:47:41.441295: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.441324: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.441340: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.441372: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.441400: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.441419: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.441468: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.441500: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.441518: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.441554: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: RandomUniform
2018-01-13 09:47:41.441586: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: VariableV2
2018-01-13 09:47:41.441604: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.441649: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.441673: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.441701: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.441724: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.441759: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.441782: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.441961: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.441990: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.442071: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.442095: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.442120: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.442142: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.443280: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.443319: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.443386: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.443411: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.443433: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.443454: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.443475: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.443495: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.443517: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1126] Converting unsupported operation: Assign
2018-01-13 09:47:41.446177: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 231 operators, 319 arrays (0 quantized)
2018-01-13 09:47:41.449197: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 1: 68 operators, 76 arrays (0 quantized)
2018-01-13 09:47:41.449990: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 2: 68 operators, 77 arrays (0 quantized)
2018-01-13 09:47:41.450763: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before dequantization graph transformations: 68 operators, 77 arrays (0 quantized)
2018-01-13 09:47:41.451432: I tensorflow/contrib/lite/toco/allocate_transient_arrays.cc:313] Total transient array allocated size: 0 bytes, theoretical optimal value: 0 bytes.

**Comment by andrehentz** [comment]
Is this a training graph? If so, you'll need to freeze it before proceeding.

**Comment by kangshan0910** [comment]
Yes, I have already frozen it.

**Comment by andrehentz** [comment]
Hi @kangshan0910 We are working on adding support for some of those ops. This is facenet, right?

**Comment by kangshan0910** [comment]
Thanks @andrehentz. This is one of the MTCNN models which include onet, pnet and rnet.

**Comment by tensorflowbutler** [comment]
The original poster has replied to this issue after the stat:awaiting response label was applied.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by andrehentz** [comment]
Quick update: Div, Exp and Sub are supported now. Neg, TensorflowMax and TensorflowSum will follow soon.

**Comment by flankechen** [comment]
@kangshan0910 Hi, I am trying to convert mtcnn to tflite as well but I got this error: https://github.com/tensorflow/tensorflow/issues/18195 
what's your progress in this issue?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 18 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Madhusakth** [comment]
Hi, 
I did almost the exact same thing to convert to a tflite file (speech model) but the output tflite file is zero bytes. Is there anything I might be missing in toco conversion?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by aselle** [comment]
Could you please try with a new nightly (pip install --upgrade tf-nightly) and let us know if that works as @andrehentz mentions, most of these ops have been added.

**Comment by kangshan0910** [comment]
Sorry for my late response, the pb file can be successfully converted now. Thank you very much for your support.

**Comment by Rob1datta** [comment]
I also had this error while converting pnet.pb to tflite ,it suggested that max is not supported by tflite
i just added --allow_custom_ops flag and the model is converted

**Comment by zh794390558** [comment]
@Rob1datta Does the model converted with --allow_custom_ops flag work well?

**Comment by jiangxiluning** [comment]
@kangshan0910 plz use my repo https://github.com/jiangxiluning/facenet_mtcnn_to_mobile

**Comment by varshadhml** [comment]
@jiangxiluning  Thank you for the github, I could convert the MTCNN model using it. 
Later, I wanted to to have variable size input for all the networks - (Pnet, Rnet and Onet). To achieve this I used tf-nightly package for conversion. 
While doing conversion I converted all the layers(cls_prob,bbox_pred,landmark_pred) one by one for Pnet, Rnet and Onet.
In the conversion step, I put "--experimental_new_converter==True".
After converting to tflite I could use dynamic size input array for all the input networks by adding one line in interpreter phase as shown below:

`from tensorflow.contrib.lite.python import interpreter`
`model = Interpreter(model_path='cls_prob.tflite')`
`input_details = model.get_input_details()`
`req_input_size = (2, 128, 200, 1)`
`model.resize_tensor_input(input_details[0]['index'], req_input_size)`
`model.allocate_tensors()`
